scala-package/core/src/main/scala/ml/dmlc/mxnet/module/BaseModule.scala

```diff
@@ -0,0 +1,552 @@
+ *    - `label_shapes`: a list of `(name, shape)`. This might be `[]` if the module does
+ *      not need labels (e.g. it does not contains a loss function at the top), or a module
+ *      is not binded for training.
+ *    - `output_shapes`: a list of `(name, shape)` for outputs of the module.
+ *
+ *  - parameters (for modules with parameters)
+ *    - `get_params()`: return a tuple `(arg_params, aux_params)`. Each of those
+ *      is a dictionary of name to `NDArray` mapping. Those `NDArray` always lives on
+ *      CPU. The actual parameters used for computing might live on other devices (GPUs),
+ *      this function will retrieve (a copy of) the latest parameters. Therefore, modifying
+ *    - `set_params(arg_params, aux_params)`: assign parameters to the devices
+ *      doing the computation.
+ *    - `init_params(...)`: a more flexible interface to assign or initialize the parameters.
+ *
+ *  - setup
+ *    - `bind()`: prepare environment for computation.
+ *    - `init_optimizer()`: install optimizer for parameter updating.
+ *
+ *  - computation
+ *    - `forward(data_batch)`: forward operation.
+ *    - `backward(out_grads=None)`: backward operation.
+ *    - `update()`: update parameters according to installed optimizer.
+ *    - `get_outputs()`: get outputs of the previous forward operation.
+ *    - `get_input_grads()`: get the gradients with respect to the inputs computed
+ *      in the previous backward operation.
+ *    - `update_metric(metric, labels)`: update performance metric for the previous forward
+ *      computed results.
+ *
+ *  - other properties (mostly for backward compatability)
+ *    - `symbol`: the underlying symbolic graph for this module (if any)
+ *      This property is not necessarily constant. For example, for `BucketingModule`,
+ *      this property is simply the *current* symbol being used. For other modules,
+ *      this value might not be well defined.
+ *
+ *  When those intermediate-level API are implemented properly, the following
+ *  high-level API will be automatically available for a module:
+ *
+ *  - `fit`: train the module parameters on a data set
+ *  - `predict`: run prediction on a data set and collect outputs
+ *  - `score`: run prediction on a data set and evaluate performance
+ * @author Yizhi Liu
+ */
+abstract class BaseModule {
+  private val logger = LoggerFactory.getLogger(classOf[BaseModule])
+
+  private[mxnet] var binded: Boolean = false
+  private[mxnet] var forTraining: Boolean = false
+  private[mxnet] var inputsNeedGrad: Boolean = false
+  private[mxnet] var paramsInitialized: Boolean = false
+  private[mxnet] var optimizerInitialized: Boolean = false
+  private[mxnet] var symbol: Symbol = null
+  private[mxnet] var execGroup: DataParallelExecutorGroup = null
+  private[mxnet] var argParams: Map[String, NDArray] = null
+  private[mxnet] var auxParams: Map[String, NDArray] = null
```

Comment:
a not-short list of package-private and mutable variables are fishy....any difficulty to make them immutable and minimize the scope?